### PR TITLE
APEXMALHAR-2414 PojoInner Join accumulation was using java reflection

### PR DESCRIPTION
**Pull request by KapoorHitesh**
@chinmaykolhatkar please review

### CODE REVIEW

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java [hunk 1]
```diff
@@ -40,6 +43,12 @@
 {
   protected final String[] keys;
   protected final Class<?> outClass;
+  private static boolean firstTimeStream;
+  private static boolean firstTimeStream2;
```
Comment (tushargosavi) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java]: you could avoid these by checking gettersStream1 null or not.

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java [hunk 2]
```diff
@@ -55,11 +64,45 @@ public PojoInnerJoin(int num, Class<?> outClass, String... keys)
...
+    for (Field field : fields) {
+      Class inputField = ClassUtils.primitiveToWrapper(field.getType());
+      String fieldName = field.getName();
+      getters.add(new KeyValPair<>(fieldName,PojoUtils.createGetter(input, fieldName, inputField)));
+    }
+    if (streamIndex == 1) {
+      gettersStream1 = getters;
+    } else {
+      gettersStream2 = getters;
+    }
```
Comment (tushargosavi) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java]: you could return the getters and assign them later. instread assign to correct variables using streamIndex. This pattern looks very common in multiple operator and creating getters and setters for multiple expressions could be made as part of PojoUtils or PojoUtilsHelpers.

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java [hunk 3]
```diff
@@ -55,11 +64,45 @@ public PojoInnerJoin(int num, Class<?> outClass, String... keys)
...
   }
 
   @Override
   public List<List<Map<String, Object>>> accumulate(List<List<Map<String, Object>>> accumulatedValue, InputT1 input)
   {
+    if (firstTimeStream) {
```
Comment (tushargosavi) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java]: check if getters1 is null and remove firstTimeStream

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java [hunk 4]
```diff
@@ -40,6 +43,12 @@
 {
   protected final String[] keys;
   protected final Class<?> outClass;
+  private static boolean firstTimeStream;
+  private static boolean firstTimeStream2;
+  private List<KeyValPair<String,PojoUtils.Getter>>  gettersStream1;
+  private List<KeyValPair<String,PojoUtils.Getter>>  gettersStream2;
```
Comment (tushargosavi) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java]: are getters serializable? if not make them transient.

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java [hunk 5]
```diff
@@ -96,27 +143,28 @@ public PojoInnerJoin(int num, Class<?> outClass, String... keys)
...
-  private Map<String, Object> pojoToMap(Object input)
+  private Map<String, Object> pojoToMap(Object input, int streamIndex)
   {
     Map<String, Object> map = new HashMap<>();
+    List<KeyValPair<String,PojoUtils.Getter>>  gettersStream;
+    if (streamIndex == 1) {
+      gettersStream = gettersStream1;
+    } else {
+      gettersStream = gettersStream2;
+    }
```
Comment (tushargosavi) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java]: could be done in conditional way
List<KeyValPair<String,PojoUtils.Getter>> gettersStream = streamIndex == 1? gettersStream1 : getterStream2;

**Comment by tushargosavi** [comment]
please take care of checkstyle errors

[ERROR] src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java:[41] (whitespace) EmptyLineSeparator: 'CLASS_DEF' should be separated from previous statement.
[ERROR] src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java:[145,79] (whitespace) WhitespaceAround: WhitespaceAround: '?' is not preceded with whitespace.

**Comment by KapoorHitesh** [comment]
@tushargosavi have done the required changes and rebased it. Please review and merge

File: library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java [hunk 6]
```diff
@@ -57,9 +64,36 @@ public PojoInnerJoin(int num, Class<?> outClass, String... keys)
...
+    return getters;
+  }
+
   @Override
   public List<List<Map<String, Object>>> accumulate(List<List<Map<String, Object>>> accumulatedValue, InputT1 input)
   {
+    if (gettersStream1 == null) {
+      gettersStream1 = createGetters(input.getClass());
+      createSetters();
```
Comment (tushargosavi) [inline library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/PojoInnerJoin.java]: setters will get create twice. can you move this check to getOutput.

**Comment by KapoorHitesh** [comment]
@tushargosavi done the required change

**Comment by KapoorHitesh** [comment]
Closing this PR as this is merged.

**Pull request closed without merging**